### PR TITLE
Provide support for hurdles

### LINKED ISSUE #255: Allow users to specify "hurdles"
[Real at al. (2020)](https://arxiv.org/abs/2003.03384) suggest that hurdles can increase the speed of an evolutionary search . This idea "[...] uses statistics of the population to early-stop the training of low performing models[...] The early-stopping criterion is the failure to reach a minimum accuracy - the _hurdle_.". Similar to their use cases a single fitness evaluation for our target applications is quite expensive, so it would make sense to adopt this strategy.

Below I suggest an implementation and, in particular, a simple API for this concept by relying on generators. This allows the user to define a fitness function as previously, but instead of `return` they will use `yield`, with each `yield` defining a "checkpoint" for early stopping. So far, this implementation only works for serial execution, a parallel version is in the making. Please let me know what you think of this in terms of usability @mschmidt87 @HenrikMettler.

```python
import numpy as np

def objective(x):
    """The object/fitness function. """
    print('first', x, '->', x)
    yield x  # first hurdle
    print('second', x, '->', x)
    yield x  # second hurdle
    print('third', x, '->', x)
    yield x + 99  # third hurdle

hurdle_rel_fitness = 0.5  # only individuals within the top 50% percentile will be allowed through each hurdle

# ==== everything below will be hidden from the user ====

 # inputs to the fitness function, i.e., individuals
combined = [0, 1, 2, 3, 4, 5, 6, 7]

# initialize generators for each individual
generators = []
for x in combined:
    generators.append(objective(x))

# keeps track of which individual still needs to be evaluated
evaluating = [True for _ in range(len(combined))]

# initial fitness evaluation until first yield, i.e., fitness after first hurdle
fitness_eval = [next(g) for g in generators]
print(fitness_eval)  # -> [0, 1, 2, 3, 4, 5, 6, 7]

while any(evaluating):

    fitness_hurdle = np.percentile([f for f in fitness_eval if f > -np.inf], hurdle_rel_fitness * 100)

    # conditional fitness evaluation; only those with fitness in upper
    # /hurdle_rel_fitness * 100/ percentile are evaluated [Real et al., 2020]
    for idx, (f, g) in enumerate(zip(fitness_eval, generators)):
        if evaluating[idx]:
            if f >= fitness_hurdle:
                try:
                    fitness_eval[idx] = next(g)
                except StopIteration:
                    evaluating[idx] = False
            else:
                fitness_eval[idx] = -np.inf
                evaluating[idx] = False

    if any(evaluating):  # only set fitness in those iterations which
                         # do not exhaust the generators; values
                         # determined in the last iteration do not
                         # represent actual fitness values
        fitness = list(fitness_eval)

print(fitness)  # -> [-inf, -inf, -inf, -inf, -inf, -inf, 105, 106]
```

### PR DESCRIPTION
**Pull request by jakobj**
This PR suggests an implementation that allows users to specify hurdles. Hurdles effectively implement early stopping for poorly performing individuals. The implementation allows users to pass lists of objectives to the EA/hl API and to define the "hurdle percentile" as a parameter of the EA for each of the objectives. Only individuals achieving fitness in the upper "hurdle percentile" of the _unique_ fitness values of the _whole_ population (parents + offspring) for the first objective are evaluated on the second objective, and so on.

An example is provided as extension of the minimal example (docstrings are yet to be adapted), where the performance difference is quite visible when considering a large number of function evaluations (28s w/o hurdles, 11s with hurdles on an i5-8350U).

Please review in particular with respect to usability and transparency of the implementation.

closes #255

### CODE REVIEW

**Comment by jakobj** [comment]
rebased on master and fixed conflicts

File: cgp/ea/mu_plus_lambda.py [hunk 1]
```diff
@@ -167,41 +169,57 @@ def _create_new_offspring_generation(self, pop: Population) -> List[IndividualBa
...
-                )
+                assert isinstance(self.process_pool, mp.pool.Pool)
+                ind_evaluating = self.process_pool.map(obj, ind_evaluating)
 
-        return sorted(combined, key=sort_func, reverse=True)
+            fitness_hurdle = compute_fitness_hurdle(ind_evaluating)
+
+            ind_evaluating_new = []
+            for ind in ind_evaluating:
+                assert isinstance(ind.fitness_current_objective, float)
```
Comment (HenrikMettler) [inline cgp/ea/mu_plus_lambda.py:211]: If I understand correctly this will raise an error if an `ind.fitness_current_objective = np.nan` - do we really want that? This can happen easily if one uses eg. division operator genes - should the run always stop then?

File: cgp/ea/mu_plus_lambda.py [hunk 2]
```diff
@@ -60,6 +61,7 @@ def __init__(
         self.local_search = local_search
         self.k_local_search = k_local_search
         self.reorder_genome = reorder_genome
+        self.hurdle_percentile = hurdle_percentile
```
Comment (HenrikMettler) [inline cgp/ea/mu_plus_lambda.py:64]: missing docstring for `hurdle_percentile`

File: cgp/individual.py [hunk 3]
```diff
@@ -28,21 +28,45 @@ class IndividualBase:
...
-    def __init__(self, fitness: Union[float, None]) -> None:
+    def __init__(self) -> None:
         """Init function.
-
-        fitness : float
-            Fitness of the individual.
         """
-        self.fitness: Union[float, None] = fitness
+        self._fitness: List[Union[None, float]] = [None]
+        self._objective_idx: int = 0
```
Comment (HenrikMettler) [inline cgp/individual.py:35]: This is a nice way of implementing it imo

File: cgp/population.py [hunk 4]
```diff
@@ -64,7 +64,7 @@ def champion(self) -> IndividualBase:
         """
 
         def key(ind: IndividualBase) -> float:
-            assert isinstance(ind.fitness, float)
+            # assert isinstance(ind.fitness, float)
```
Comment (HenrikMettler) [inline cgp/population.py:67]: remove

**Review by HenrikMettler — CHANGES_REQUESTED** [review]

File: cgp/ea/mu_plus_lambda.py [hunk 5]
```diff
@@ -26,6 +26,7 @@ def __init__(
         local_search: Callable[[IndividualBase], None] = lambda combined: None,
         k_local_search: Union[int, None] = None,
         reorder_genome: bool = False,
+        hurdle_percentile: List = [0.0]
```
[the remark this answers is not in the document]
Comment (jakobj) [reply cgp/ea/mu_plus_lambda.py:29]: the issue with an additional argument would be redundancy: what if `use_hurdles` is `True`, but `hurdle_percentile` is `[0.0]` or `[0.0, 0.0]`? should this be an error? a warning? the way it is handled now is easier for the user i think

File: cgp/ea/mu_plus_lambda.py (hunk 1 above)
Comment (jakobj) [reply cgp/ea/mu_plus_lambda.py:211]: nope, this wouldn't raise an error:
```python
In [3]: isinstance(np.nan, float)
Out[3]: True
```

**Comment by jakobj** [comment]
thanks for the review @HenrikMettler, i've implemented your suggestions now

File: examples/example_hurdles.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
+"""
+Minimal example for evolutionary regression
```
Comment (HenrikMettler) [inline examples/example_hurdles.py:2]: please adapt the docstrings before merging

File: examples/example_minimal.py [hunk 7]
```diff
@@ -42,10 +42,11 @@ def f_target(x):
 
 
 def objective(individual):
-    if individual.fitness is not None:
+
```
Comment (HenrikMettler) [inline examples/example_minimal.py:45]: Revert these changes (?)

**Review by HenrikMettler — APPROVED** [review]
Please just adapt the new example before merging

**Comment by jakobj** [comment]
thanks @HenrikMettler ! implemented your suggestions, now waiting for travis
